File: daily_python/33_Search_in_Rotated_Sorted_Array.py
import os
import sys
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l, r = 0, n - 1
        while l < r:
            m = (l + r) // 2
            if nums[m] < nums[r]:
                r = m
            else:
                l = m + 1
        d = r
        l, r = 0, n - 1
        while l <= r:
            m = (l + r) // 2
            if nums[(m + d) % n] == target:
                return (m + d) % n
            if nums[(m + d) % n] < target:
                l = m + 1
            else:
                r = m - 1
        return -1
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            target = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((nums, target))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, target in dataset:
        results.append(solution.search(nums, target))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2026/33_Search_in_Rotated_Sorted_Array.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0;
        int right = n - 1;
        while(left < right){
            int mid = left + (right - left) / 2;
            if (nums[mid] < nums[right]) right = mid;
            else left = mid + 1;
        }
        int start = right;
        int l = 0;
        int r = n - 1;
        while(l <= r){
            int m = l + (r - l) / 2;
            if (nums[(m + start) % n] == target) return (m + start) % n;
            if (nums[(m + start) % n] < target) l = m + 1;
            else r = m - 1;
        }
        return -1;
    }
};

int main(){
    vector<int> nums = {4,5,6,7,0,1,2};
    int target = 0;
    // vector<int> nums = {4,5,6,7,0,1,2};
    // int target = 3;
    // vector<int> nums = {1};
    // int target = 0;
    Solution S;
    cout << S.search(nums, target) << endl;
    return 0;
}